server/models/procedures/voucher.sql

```diff
@@ -220,4 +220,80 @@ BEGIN
   END IF;
 END $$
 
+
+-- this is handled on change rate changed
+DROP PROCEDURE IF EXISTS `UpdateLossAccount`$$
+CREATE   PROCEDURE `UpdateLossAccount`(
+  IN _enterpriseId INT,
+  IN _project_id INT, 
+  IN _user_id INT,
+  IN _currencyId INT,
+  IN _date DATE, 
+  IN _rate DECIMAL(19, 4)
+  )
+BEGIN
+DECLARE _account_id, _loss_account_id, _gain_account_id INT(10) UNSIGNED;
+DECLARE _privousRate, _balance, _difference DECIMAL(19, 4);
+DECLARE _fiscalyearID, _periodID  mediumint(8) UNSIGNED;
+DECLARE voucherUUID BINARY(16);
+DECLARE _description, _account_label VARCHAR(255);
+
+DECLARE done BOOLEAN;
+DECLARE curs1 CURSOR FOR 
+  SELECT c.account_id , CONCAT(a.number , ' - ', a.label) 
+  FROM cash_box_account_currency c
+  JOIN account a ON a.id =  c.account_id
+  WHERE currency_id = _currencyId;
+   
+DECLARE CONTINUE HANDLER FOR NOT FOUND SET done = TRUE;
+
+SET _privousRate = (SELECT GetExchangeRate(_enterpriseId, _currencyId, _date));
+
+SET _fiscalyearID = (SELECT f.id FROM fiscal_year f WHERE f.start_date <= DATE(_date) AND DATE(_date) <= f.end_date);
+
+SET _periodID = (SELECT p.id FROM period p WHERE p.start_date <= DATE(_date) AND DATE(_date) <= p.end_date);
+
+SET _loss_account_id = (SELECT loss_account_id FROM enterprise WHERE id = _enterpriseId);
+SET _gain_account_id = (SELECT gain_account_id FROM enterprise Where id= _enterpriseId);
+
+SET _description = "Changement taux d'echange(exchange rate updated)";
+
+OPEN curs1;
+    read_loop: LOOP
+    
+    FETCH curs1 INTO _account_id, _account_label;
+    
+    IF done THEN
+      LEAVE read_loop;
+    END IF;
+        
+    SET _balance = ( 
+		 	SELECT (debit - credit) 
+			FROM period_total 
+			WHERE account_id = _account_id and fiscal_year_id= _fiscalyearID);
+		
+		SET _difference = (_rate*_balance) - (_privousRate*_balance);
+		
+		IF _balance <> NULL THEN
+			SET voucherUUID = HUID(uuid());
+			
+			INSERT INTO voucher(uuid, date, project_id, currency_id, amount, description, user_id, type_id)
+			VALUES(voucherUUID, _date, _project_id, _currencyId, ABS(_difference), CONCAT(_description,', ', _account_label), _user_id, 18);
+			
+			IF (_difference > 0) THEN
+				INSERT INTO voucher_item(uuid, account_id, debit, credit,voucher_uuid)
+				VALUES(HUID(uuid()), _loss_account_id, ABS(_difference), 0, voucherUUID);	
+			ELSE
+				INSERT INTO voucher_item(uuid, account_id, debit, credit,voucher_uuid)
+				VALUES(HUID(uuid()), _gain_account_id, ABS(_difference), 0, voucherUUID);
```

Comment:
In the case of gain, we must credit the gain account.

In the case of loss we have to debit the loss account as you have done